// sentence-splitter/lib/sentence-splitter.js
// LICENSE : MIT
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Syntax = undefined;
exports.default = splitSentences;
exports.createWhiteSpaceNode = createWhiteSpaceNode;
exports.createSentenceNode = createSentenceNode;

var _structuredSource = require("structured-source");

var _structuredSource2 = _interopRequireDefault(_structuredSource);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultOptions = {
    charRegExp: /[\.。\?\!？！]/,
    newLineCharacters: "\n"
};
var Syntax = exports.Syntax = {
    "WhiteSpace": "WhiteSpace",
    "Sentence": "Sentence"
};
function splitSentences(text) {
    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var matchChar = options.charRegExp || defaultOptions.charRegExp;
    var newLineCharacters = options.newLineCharacters || defaultOptions.newLineCharacters;
    var src = new _structuredSource2.default(text);
    var createNode = function createNode(type, start, end) {
        var range = [start, end];
        var location = src.rangeToLocation(range);
        var slicedText = text.slice(start, end);
        var node = undefined;
        if (type === Syntax.WhiteSpace) {
            node = createWhiteSpaceNode(slicedText, location, range);
        } else if (type === Syntax.Sentence) {
            node = createSentenceNode(slicedText, location, range);
        }
        return node;
    };
    var results = [];
    var startPoint = 0;
    var currentIndex = 0;
    var isSplitPoint = false;
    var newLineCharactersLength = newLineCharacters.length;
    for (; currentIndex < text.length; currentIndex++) {
        var char = text[currentIndex];
        var whiteTarget = text.slice(currentIndex, currentIndex + newLineCharactersLength);
        if (whiteTarget === newLineCharacters) {
            // (string)\n
            if (startPoint !== currentIndex) {
                results.push(createNode(Syntax.Sentence, startPoint, currentIndex));
            }
            for (var i = 0; i < newLineCharactersLength; i++) {
                // string(\n)
                var startIndex = currentIndex + i;
                results.push(createNode(Syntax.WhiteSpace, startIndex, startIndex + 1));
            }
            // string\n|
            startPoint = currentIndex + newLineCharactersLength;
            isSplitPoint = false;
        } else if (matchChar.test(char)) {
            isSplitPoint = true;
        } else {
            // why `else`
            // it for support 。。。 pattern
            if (isSplitPoint) {
                results.push(createNode(Syntax.Sentence, startPoint, currentIndex));
                // reset stat
                startPoint = currentIndex;
                isSplitPoint = false;
            }
        }
    }

    if (startPoint !== currentIndex) {
        results.push(createNode(Syntax.Sentence, startPoint, currentIndex));
    }
    return results;
}
function createWhiteSpaceNode(text, loc, range) {
    return {
        type: "WhiteSpace",
        raw: text,
        value: text,
        loc: loc,
        range: range
    };
}
function createSentenceNode(text, loc, range) {
    return {
        type: "Sentence",
        raw: text,
        value: text,
        loc: loc,
        range: range
    };
}
