// textlint-rule-preset-jtf-style/lib/2.1.5.js
// LICENSE : MIT
"use strict";
/*
2.1.5.カタカナ
カタカナは「全角」で表記します。
半角カタカナは特殊な用途を除いて、原則として使いません。

Halfwidth Katakana variants（半角片仮名）
\uFF65-\uFF9F とする
http://www.asahi-net.or.jp/~ax2s-kmtn/ref/unicode/uff00.html
 */

var _nodeUtil = require("./util/node-util");

var _regexp = require("./util/regexp");

var _textlintRulePrh = _interopRequireDefault(require("textlint-rule-prh"));

var _path = _interopRequireDefault(require("path"));

var _matchIndex = require("match-index");

var _moji = _interopRequireDefault(require("moji"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 半角カタカナを全角カタカナに変換
 *
 * @param {String} str 変換したい文字列
 */
function toZenkaku(string) {
  return (0, _moji.default)(string).convert("HK", "ZK").toString();
}

function reporter(context) {
  var Syntax = context.Syntax,
      RuleError = context.RuleError,
      fixer = context.fixer,
      report = context.report,
      getSource = context.getSource; // 辞書ベースのカタカタ表記のチェックを行う

  var dictRule = _textlintRulePrh.default.fixer(context, {
    rulePaths: [_path.default.join(__dirname, "..", "dict", "2.1.5.yml")]
  });

  var originalStrRule = dictRule[Syntax.Str]; // 半角カタカナの使用をチェックする

  dictRule[Syntax.Str] = function (node) {
    originalStrRule(node);

    if (!(0, _nodeUtil.isUserWrittenNode)(node, context)) {
      return;
    }

    var text = getSource(node);
    var matches = (0, _matchIndex.matchCaptureGroupAll)(text, /([\uFF65-\uFF9F]+)/g);
    matches.forEach(function (match) {
      var index = match.index,
          text = match.text;
      report(node, new RuleError("カタカナは「全角」で表記します。", {
        index: index,
        fix: fixer.replaceTextRange([index, index + text.length], toZenkaku(text))
      }));
    });
  };

  return dictRule;
}

module.exports = {
  linter: reporter,
  fixer: reporter
};
