// prh/lib/target.js
"use strict";
var r = require("./utils/regexp");
var targetPattern_1 = require("./targetPattern");
var Target = (function () {
    function Target(src) {
        if (!src) {
            throw new Error("src is requried");
        }
        this.file = r.parseRegExpString(src.file);
        if (!this.file) {
            this.file = new RegExp(r.escapeSpecialChars(src.file));
        }
        if (src.includes) {
            this.includes = src.includes.map(function (include) { return new targetPattern_1.default(include); });
        }
        else {
            this.includes = [];
        }
        if (src.excludes) {
            this.excludes = src.excludes.map(function (exclude) { return new targetPattern_1.default(exclude); });
        }
        else {
            this.excludes = [];
        }
    }
    Target.prototype.reset = function () {
        this.file.lastIndex = 0;
        this.includes.forEach(function (include) { return include.reset(); });
        this.excludes.forEach(function (exclude) { return exclude.reset(); });
    };
    Target.prototype.toJSON = function () {
        var alt = {};
        for (var key in this) {
            if (key.indexOf("_") === 0) {
                continue;
            }
            var value = this[key];
            if (value instanceof RegExp) {
                alt[key] = value.toString();
                continue;
            }
            alt[key] = value;
        }
        return alt;
    };
    return Target;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Target;
