// prh/lib/changeset/diff.js
"use strict";
var Diff = (function () {
    function Diff(pattern, expected, index, matches, rule) {
        this.pattern = pattern;
        this.expected = expected;
        this.index = index;
        this.matches = matches;
        this.rule = rule;
    }
    Object.defineProperty(Diff.prototype, "tailIndex", {
        get: function () {
            return this.index + this.matches[0].length;
        },
        enumerable: true,
        configurable: true
    });
    Diff.prototype.isEncloser = function (other) {
        return this.index < other.index && other.tailIndex < this.tailIndex;
    };
    Diff.prototype.isCollide = function (other) {
        if (other.index < this.index && this.index < other.tailIndex) {
            return true;
        }
        if (this.index < other.index && other.index < this.tailIndex) {
            return true;
        }
        return false;
    };
    Diff.prototype.isBefore = function (other) {
        return this.index < other.index;
    };
    return Diff;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Diff;
